plugins/inputs/influxdb_listener/influxdb_listener.go

```diff
@@ -0,0 +1,395 @@
+	return sampleConfig
+}
+
+func (h *InfluxDBListener) Description() string {
+	return "InfluxDB listener"
+}
+
+func (h *InfluxDBListener) Gather(_ telegraf.Accumulator) error {
+	return nil
+}
+
+func (h *InfluxDBListener) routes() {
+	authHandler := internal.AuthHandler(h.BasicUsername, h.BasicPassword, "influxdb",
+		func(_ http.ResponseWriter) {
+			h.authFailures.Incr(1)
+		},
+	)
+
+	h.mux.Handle("/write", authHandler(h.handleWrite()))
+	h.mux.Handle("/query", authHandler(h.handleQuery()))
+	h.mux.Handle("/ping", h.handlePing())
+	h.mux.Handle("/", authHandler(h.handleDefault()))
+}
+
+func (h *InfluxDBListener) Init() error {
+	tags := map[string]string{
+		"address": h.ServiceAddress,
+	}
+	h.bytesRecv = selfstat.Register("influxdb_listener", "bytes_received", tags)
+	h.requestsServed = selfstat.Register("influxdb_listener", "requests_served", tags)
+	h.writesServed = selfstat.Register("influxdb_listener", "writes_served", tags)
+	h.queriesServed = selfstat.Register("influxdb_listener", "queries_served", tags)
+	h.pingsServed = selfstat.Register("influxdb_listener", "pings_served", tags)
+	h.requestsRecv = selfstat.Register("influxdb_listener", "requests_received", tags)
+	h.notFoundsServed = selfstat.Register("influxdb_listener", "not_founds_served", tags)
+	h.buffersCreated = selfstat.Register("influxdb_listener", "buffers_created", tags)
+	h.authFailures = selfstat.Register("influxdb_listener", "auth_failures", tags)
+	h.routes()
+
+	if h.MaxBodySize.Size == 0 {
+		h.MaxBodySize.Size = defaultMaxBodySize
+	}
+
+	if h.ReadTimeout.Duration < time.Second {
+		h.ReadTimeout.Duration = time.Second * 10
+	}
+	if h.WriteTimeout.Duration < time.Second {
+		h.WriteTimeout.Duration = time.Second * 10
+	}
+
+	return nil
+}
+
+// Start starts the InfluxDB listener service.
+func (h *InfluxDBListener) Start(acc telegraf.Accumulator) error {
+	h.acc = acc
+
+	tlsConf, err := h.ServerConfig.TLSConfig()
+	if err != nil {
+		return err
+	}
+
+	h.server = http.Server{
+		Addr:         h.ServiceAddress,
+		Handler:      h,
+		ReadTimeout:  h.ReadTimeout.Duration,
+		WriteTimeout: h.WriteTimeout.Duration,
+		TLSConfig:    tlsConf,
+	}
+
+	var listener net.Listener
+	if tlsConf != nil {
+		listener, err = tls.Listen("tcp", h.ServiceAddress, tlsConf)
+		if err != nil {
+			return err
+		}
+	} else {
+		listener, err = net.Listen("tcp", h.ServiceAddress)
+		if err != nil {
+			return err
+		}
+	}
+	h.listener = listener
+	h.port = listener.Addr().(*net.TCPAddr).Port
+
+	h.wg.Add(1)
+	go func() {
+		defer h.wg.Done()
+		h.server.Serve(h.listener)
+	}()
+
+	h.log.Infof("Started HTTP listener service on %s", h.ServiceAddress)
+
+	return nil
+}
+
+// Stop cleans up all resources
+func (h *InfluxDBListener) Stop() {
+	deadline := time.Now().Add(10 * time.Second)
+	ctx, cancel := context.WithDeadline(context.Background(), deadline)
+	defer cancel() // go vet requires us to call cancel even though we're waiting for the deadline
```

Comment:
what happens if the deadline passes, will wait hang?

Reply:
This change is what we think will fix the panics mentioned in influxdata/EAR#1379.  They seem to be caused by http server connections receiving data after the server is shutdown.  Before this change we would close the listeners directly but not shut down existing connections.

We think using Server.Shutdown will fix the problem because in addition to closing listeners it also shuts down existing connections.  We didn't do that before because Server.Shutdown didn't exist at the time.  The last change to this file related to server shutdown was Oct 2016 and Server.Shutdown was added in go 1.8 in Feb 2017.  

To answer your question, I don't think wait will hang if the deadline passes because Server.Serve should return as soon as the listeners are closed.  But now that I review it more closely I think it's possible that this won't completely prevent the condition that causes the panics.  If the connections don't close in the 10 seconds the panics could still happen.

I don't think we want to wait indefinitely but the deadline may just set up a race condition.  What do you think we should do?  @danielnelson?

Reply:
Wait will wait until Server returns (above in Connect).  I assumed that the Server waited for all handler goroutines before returning from Serve, but I guess I'm not 100% sure.  You could probably check for this by temporarily adding a Sleep call into the handler and seeing how long it takes to complete.

Nitpick: Since a deadline is absolute but a timeout is relative, I'd use a timeout so you don't need to add it to time.Now.

Reply:
👍 WithTimeout.

> it also shuts down existing connections.

To be pedantic, it stops listening and lets the connections finish what they're doing. It's still possible for them to write a response before the deadline, and it's possible that they don't finish at all before the server exits. A good practice for long-running http connections would be to pass a main context down through the app stack so that requests can "notice" if the app is trying to shut down (their contexts are cancelled), and gracefully stop doing work.
details here https://golang.org/src/net/http/server.go?s=83923:83977#L2643

> the deadline may just set up a race condition

I don't think that's the case, other than the aforementioned requests-not-finishing-in-time, which is fine.

Reply:
I reworked this to have the connection pay attention to the context.  With connections closing quickly we can let shutdown happen gracefully and remove the 10 second timeout.  Shutdown will block which prevents the accumulator from going away while the connection needs it, so the panic should still be prevented.